design-proposals/nic-hotplug/nic-hotplug.md

```diff
@@ -0,0 +1,234 @@
+
+## Non Goals
+- Providing multiple connections to the primary cluster network (managed by
+Kubernetes).
+- Dynamic attachment / removal of SR-IOV networks; since these rely on a device
+plugin - and thus require the update of the pod's resources - i.e. request the
+VF resource - something which, is not possible on a running pod. This particular
+goal requires more study, and should be discussed in a separate design document.
+
+## Definition of Users
+This feature is intended for cluster users who want to connect their existing
+Virtual Machines to additional existing networks.
+These users may - or may not - have permissions to create the aforementioned
+additional networks; if so, they would need to rely on the cluster
+administrator to provision it for them.
+
+## User Stories
+* as a user, I want to add a new network interface to a running VM.
+* as a user, I want to remove a new network interface from a running VM.
+
+## Repos
+- [CNAO](https://github.com/kubevirt/cluster-network-addons-operator)
+- [KubeVirt](https://github.com/kubevirt/kubevirt)
+- [Multus](https://github.com/k8snetworkplumbingwg/multus-cni)
+
+# Design
+
+## Multus
+Multus - a CNI plugin - only handles the ADD / REMOVE verb, and is triggered
+by kubelet only when the pod's sandbox is created - or removed. Given its
+simplicity, it assumes no networks exist whenever it is executed, and procceeds
+to call the ADD / DEL for **all** networks listed in its
+`k8s.v1.cni.cncf.io/networks` annotation.
+
+As such, the first part of the design implementation must focus on multus; it
+must be refactored to work as a thick plugin, which enables it to be triggered
+not only when the pod's sandbox is created, but also on-demand.
+
+### Thickening multus
+A "thin CNI plugin" runs as a one-shot process, typically as a binary on disk
+executed on a Kubernetes host machine.
+
+A "thick CNI Plugin", on the other hand, is a CNI component composed of two (or
+more) parts, usually composed of "shim", and a long lived process (daemon)
+resident in memory. The "shim" is a lightweight "thin CNI plugin" component that
+simply passes CNI parameters (such as JSON configuration, and environment
+variables) to the daemon component, which then processes the CNI request.
+
+To transform multus into a thick plugin, it is needed to instantiate a long
+lived process - which will be the multus pod entrypoint - listening to a unix
+domain socket - this socket must be available both in the multus pod and the
+hosts's mount namespaces; as such, a bind mount to host this socket must be
+provided for the multus pod.
+
+The CNI shim will then be invoked by kubelet, and send the CNI ADD/DELETE
+commands to the server side of multus via the unix domain socket previously
+mentioned.
+
+Refer to the sequence diagram below for more information:
+
+![Create pod flow for thick plugin](create-pod-flow.png)
+
+### Further multus requirements
```

Comment:
From current multus user's point of view, we should keep current multus binary and add thick plugin in another one. As Doug mention, if we make thick plugin in same binary (with two 'mode') then current multus binary gets bigger than current. Multus is also used in edge case which requires small footprint and its user wants to use current thin plugin and they wants to keep current binary size.

So could you please add following into the requirements?

- keep current binary as thin plugin and thick plugin will be another binary
- try to keep current binary size for thin plugin (no new feature for thin plugin, well we may need a few code changes)
- multus provides both daemonset (thick plugin and thin plugin)

Reply:
OK, I get it - thanks. Makes sense.

IIUC, we'd have to ship both executables on the docker images, and have two types of multus daemonset specs, each with different entrypoint commands. Right ?

Reply:
Right, two daemonset (multus-daemonset.yml and multus-daemonset-thick.yml) and one container image which contains both binary.

Reply:
Done.

Please let me know if the info I've added is clear enough.